RGB.NET.Devices.WS281X/Sketches/RGB.NET_udp.ino

```diff
@@ -0,0 +1,210 @@
+  {
+    lastSequenceNumbers[i] = 255;
+  }
+   
+  WiFi.begin(ssid, password); 
+  while (WiFi.status() != WL_CONNECTED)
+  {
+    delay(500);
+  }  
+  delay(100);
+  
+  Udp.begin(PORT);
+  server.begin();
+}
+
+bool checkSequenceNumber(int channel, byte currentSequenceNumber)
+{
+  bool isValid = (currentSequenceNumber > lastSequenceNumbers[channel]) || ((lastSequenceNumbers[channel] > 200) && (currentSequenceNumber < 50));
+  if(isValid)
+  {
+    lastSequenceNumbers[channel] = currentSequenceNumber;
+  }
+  return isValid;
+}
+
+void loop() {
+  // Web client
+  WiFiClient client = server.available();
+  if (client) 
+  {
+    String currentLine = "";
+    while (client.connected())
+    {
+      if (client.available())
+      {
+        char c = client.read();
+        header += c;
+        if (c == '\n') {
+          if (currentLine.length() == 0) {
+            client.println("HTTP/1.1 200 OK");
+            client.println("Content-type:text/html");
+            client.println("Connection: close");
+            client.println();
+            
+            if (header.indexOf("GET /reset") >= 0)
+            {
+              for(int i = 0; i < CHANNELS; i++)
+              {
+                lastSequenceNumbers[i] = 255;                
+              }
+              for(int i = 0; i < LEDS_CHANNEL_1; i++)
+              {
+                leds_channel_1[i] = CRGB::Black;               
+              }
+              for(int i = 0; i < LEDS_CHANNEL_2; i++)
+              {
+                leds_channel_2[i] = CRGB::Black;               
+              }
+              for(int i = 0; i < LEDS_CHANNEL_3; i++)
+              {
+                leds_channel_3[i] = CRGB::Black;               
+              }
+              for(int i = 0; i < LEDS_CHANNEL_4; i++)
+              {
+                leds_channel_4[i] = CRGB::Black;               
+              }
+              FastLED.show();
+            }
+            else if (header.indexOf("GET /channels") >= 0)
+            {
+              client.println(CHANNELS);
+            }
+            else if (header.indexOf("GET /channel/1") >= 0)
+            {
+              client.println(LEDS_CHANNEL_1);
+            }
+            else if (header.indexOf("GET /channel/2") >= 0)
+            {
+              client.println(LEDS_CHANNEL_2);
+            }
+            else if (header.indexOf("GET /channel/3") >= 0)
+            {
+              client.println(LEDS_CHANNEL_3);
+            }
+            else if (header.indexOf("GET /channel/4") >= 0)
+            {
+              client.println(LEDS_CHANNEL_4);
+            }
+            client.println();
+            
+            break;
+          }
+          else 
+          {
+            currentLine = "";
+          }
+        }
+        else if (c != '\r') 
+        {
+          currentLine += c;
+        }
+      }
+    }
+    header = "";
+    client.stop();
+  }
+
+  // Color update
+  int packetSize = Udp.parsePacket();
```

Comment:
Why is the color update not also done via the HTTP server?

Reply:
The achievable update rates are way higher like this. I might extend it though to be usable with pure http.

Reply:
What about using `Connection: keep-alive` ? Also it would be easier to use a higher level abstraction of the http handling by using `ESP8266WebServer` instead of implementing the low level http connection by yourself.

Reply:
I don't even consider computation related things here. The main point for me is the fact, that the RTT with a wireless connection is quite bad. I get 2-3ms with a lot of peaks with 10-20ms with super short distance and no other devices in the wlan, which is already quite a lot. I guess that with a average wlan the 10-20 is the average and the peaks are even higher.
Using udp my latency is only one way (there's no round trip) and if a package is delayed by such a peak it's just dropped (if the next one arrives before it).
Using http (and therefore tcp) I have that overhead of acknowledgement and so on. Also with the bigger header it increases the change of package-fragmentation (which isn't a problem with tcp, but also increases the transmission time).

I'll add a full rest-api so it can be used with pure http, but in my opinion in most cases the udp transfer is preferable.

Reply:
Updating over http is even worse than expected. Not sure if I did something wrong here, but it takes the http-server ~250ms to handle the post-request. (That doesn't include the logic I'm doing - which is cheaper than I expected. It's only measured from the start of the loop function to the start of the handleUpdate function.)
For real-time applications this is absolute unusable but it might be fine if not fast but reliable updates are required.